File: chap08-image.py


import os
os.chdir("D:\\nonstruct")


################################
##  图像处理                  ##
################################

from PIL import Image, ImageFilter
lena = Image.open("lena.jpg")
lena.size
lena.show()

# 和数组的转换
import numpy as np
d1 = np.asarray(lena)
img1 = Image.fromarray(d1)

# 图像转换
img2 = img1.resize((200,100))
img3 = img1.rotate(90) # 逆时针旋转90
img4 = img1.transpose(Image.FLIP_LEFT_RIGHT) # 左右翻转
img5 = img1.transpose(Image.FLIP_TOP_BOTTOM) # 上下翻转

img6 = img1.crop((0, 0, 200, 100)) #左上右下

# 矩阵运算
lena1 = Image.fromarray(d1 + 100)
lena2 = Image.fromarray(d1*3)

# 卷积
con1 = ImageFilter.Kernel((3, 3), (4,0,0,0,0,0,0,0,-4), scale = 9)
img7 = lena.filter(con1)





################################
##  使用图像数据进行手写识别  ##
################################

import mxnet as mx
import os
import re
from PIL import Image
import numpy as np

fn = os.listdir("MINIST/train")
y = [eval(re.sub("(^[0-9]*_|\\.jpg$)", "", x)) for x in fn]
y = mx.nd.array(y)

X = mx.nd.zeros((len(fn), 1, 32, 32))
for i in range(len(fn)):
    img0 = Image.open("MINIST/train/" + fn[i])
    img0 = img0.resize((32, 32))
    arr1 = np.asarray(img0)[:,:,1]
    X[i][0] = arr1 / 255


fn = os.listdir("MINIST/test")
y1 = [eval(re.sub("(^[0-9]*_|\\.jpg$)", "", x)) for x in fn]
y1 = mx.nd.array(y1)

X1 = mx.nd.zeros((len(fn), 1, 32, 32))
for i in range(len(fn)):
    img0 = Image.open("MINIST/test/" + fn[i])
    img0 = img0.resize((32, 32))
    arr1 = np.asarray(img0)[:,:,1]
    X1[i][0] = arr1 / 255   


batch_size = 100
train_iter = mx.io.NDArrayIter(X, y, batch_size, shuffle = True)
val_iter = mx.io.NDArrayIter(X1, y1, batch_size, shuffle = True)

data = mx.sym.var("data")
C1 = mx.sym.Convolution(data = data, kernel = (5,5), num_filter = 20)
tanh1 = mx.sym.Activation(data = C1, act_type = "tanh")
S2 = mx.sym.Pooling(data = tanh1, pool_type = "max", kernel = (2, 2), stride = (2, 2))
C3 = mx.sym.Convolution(data = S2, kernel = (5, 5), num_filter = 50)
tanh2 = mx.sym.Activation(data = C3, act_type = "tanh")
S4 = mx.sym.Pooling(data = tanh2, pool_type = "max", kernel = (2, 2), stride = (2, 2))
flatten = mx.sym.flatten(data = S4)
C5 = mx.symbol.FullyConnected(data = flatten, num_hidden = 120)
tanh3 = mx.sym.Activation(data = C5, act_type = "tanh")
F6 = mx.sym.FullyConnected(data = tanh3, num_hidden = 84)
tanh4 = mx.sym.Activation(data = F6, act_type = "tanh")
F7 = mx.sym.FullyConnected(data = tanh4, num_hidden = 10)
lenet = mx.sym.SoftmaxOutput(data = F7, name = "softmax")

ctx = mx.cpu()
import datetime
import logging
logging.getLogger().setLevel(logging.DEBUG) 

m1 = mx.mod.Module(symbol = lenet, context = ctx)

mx.random.seed(0)
t1 = datetime.datetime.now()
m1.fit(train_iter, eval_data = val_iter, optimizer = "sgd", optimizer_params = {'learning_rate':0.05,'momentum': 0.9}, eval_metric = 'acc', num_epoch = 100)
t2 = datetime.datetime.now()
t2 - t1

# datetime.timedelta(seconds=708, microseconds=917126)
# datetime.timedelta(seconds=36, microseconds=231106)

prob1 = m1.predict(val_iter)
p1 = prob1.argmax(axis = 1)
np.mean(p1.asnumpy() == y1.asnumpy())

File: chap08-image.R


setwd("D:\\nonstruct")


################################
##  图像处理                  ##
################################

library(EBImage)
lena <- readImage("lena.jpg")
dim(lena)
display(lena)

# 和数组的转换
d1 <- as.array(lena)
dim(d1)
img1 <- Image(d1[, , 1])

# 图像转换
img2 <- resize(img1, 200, 100)
img3 <- rotate(img1, 90) # 顺时针旋转90
img4 <- flop (img1) # 左右翻转
img5 <- flip (img1) # 上下翻转

img6 <- img1[1:200, 1:100] #左到右、上到下

# 矩阵运算
lena1 <- lena + 0.5
lena2 <- 3 * lena

# 卷积
con1 <- matrix(c(4,0,0,0,0,0,0,0,-4), 3, 3)
img7 <- filter2(img1, con1)




################################
##  使用图像数据进行手写识别  ##
################################

setwd("D:\\Storage\\Nutstore\\Book\\rucds\\work\\nonstruct")
library(mxnet)
library(EBImage)

fn <- list.files("MINIST/train")
y <- as.numeric(gsub("(^[0-9]*_|\\.jpg$)", "", fn))
X <- array(0, dim = c(32, 32, 1, length(fn)))
for (i in seq_along(fn)) {
	img0 <- readImage(file.path("MINIST/train", fn[i]))
	img0 <- resize(img0, 32, 32)
	arr1 <- as.array(img0)[, , 1]
	X[, , 1, i] <- arr1
}

fn <- list.files("MINIST/test")
y1 <- as.numeric(gsub("(^[0-9]*_|\\.jpg$)", "", fn))
X1 <- array(0, dim = c(32, 32, 1, length(fn)))
for (i in seq_along(fn)) {
	img0 <- readImage(file.path("MINIST/test", fn[i]))
	img0 <- resize(img0, 32, 32)
	arr1 <- as.array(img0)[, , 1]
	X1[, , 1, i] <- arr1
}

train_iter = list(data = X, label = y)
val_iter = list(data = X1, label = y1)

# 建立 CNN 模型
data <- mx.symbol.Variable("data")
C1 <- mx.symbol.Convolution(data = data, kernel = c(5,5), num_filter = 20)
tanh1 <- mx.symbol.Activation(data = C1, act_type = "tanh")
S2 <- mx.symbol.Pooling(data  = tanh1, pool_type = "max", kernel = c(2, 2), stride = c(2, 2))
C3 <- mx.symbol.Convolution(data = S2, kernel = c(5, 5), num_filter = 50)
tanh2 <- mx.symbol.Activation(data = C3, act_type = "tanh")
S4 <- mx.symbol.Pooling(data = tanh2, pool_type = "max", kernel = c(2, 2), stride = c(2, 2))
flatten <- mx.symbol.Flatten(data = S4)
C5 <- mx.symbol.FullyConnected(data = flatten, num_hidden = 120)
tanh3 <- mx.symbol.Activation(data = C5, act_type = "tanh")
F6 <- mx.symbol.FullyConnected(data = tanh3, num_hidden = 84)
tanh4 <- mx.symbol.Activation(data = F6, act_type = "tanh")
F7 <- mx.symbol.FullyConnected(data = tanh4, num_hidden = 10)
lenet <- mx.symbol.SoftmaxOutput(data = F7, name = "softmax")


# 可视化
graph.viz(lenet, direction = "LR")

ctx = mx.gpu()

# 基于 CPU 运行模型
mx.set.seed(0)
t1 <- Sys.time()
m1 <- mx.model.FeedForward.create(lenet, X = X, y = y, ctx = ctx, num.round = 100, array.batch.size = 100, learning.rate = 0.05, momentum=  0.9, eval.metric = mx.metric.accuracy, eval.data = val_iter)
t2 <- Sys.time()
t2 - t1

# Time difference of 10.83941 mins
# Time difference of 35.35645 secs

# 预测
prob1 <- predict(m1, X = X1)
p1 <- apply(prob1, 2, which.max) - 1
sum(p1 == y1) / length(p1)
table(p1, y1)



